### DISCUSSION TITLE
Virtual GamePad Layout Help

### DISCUSSION THREAD

**Discussion by S-Coding23** [EmulatorJS/EmulatorJS#544] [Q&A]
I have noticed that the virtual gamepad on iOS and Android for NES games (So far I have only tested NES) has the "A" and "B" buttons reversed. They should be "B" on the left then "A" on the right. I have tried following the documentation for setting a custom layout, but have found that using either one of the Zone or DPad options causes the emulator to not load at all. 

 {
        type: "zone",
        location: "left",
        left: "50%",
        top: "50%",
        joystickInput: true,
        inputValues: [16, 17, 18, 19]
    },
    {
        type: "dpad"
        location: "left",
        left: "50%",
        right: "50%",
        joystickInput: false,
        inputValues: [4, 5, 6, 7]
    },

If I leave those options out, the emulator loads, but doesnt have a default dpad or joystick, so that doesnt work. 

I edited the gamepad settings in emu-main.js but this only fixed the layout for iOS and not Android. 

Also iOS doesnt have an option to go full screen but Android does. 

Any assistance on these issues would be greatly appreciated.

**Comment by ethanaobrien** [comment]
>If I leave those options out, the emulator loads, but doesnt have a default dpad or joystick, so that doesnt work.

Could you send a log of the console when it doesn't work?

>I edited the gamepad settings in emu-main.js but this only fixed the layout for iOS and not Android.

Did you minify the files? Minified files aren't pulled on ios and are only used on android/every other device.

>Also iOS doesnt have an option to go full screen but Android does.

That's a Safari thing and it cannot be fixed. Cheers to the most incompatible and overly-used browser known to man!

**Reply by S-Coding23** [reply]
I should be able to get the logs sometime tomorrow. 
I did not originally minify the files but I just did and that resolved the layout on the Android. Thank you!!
Understood about Safari :)

**Comment by S-Coding23** [comment]
I attached the console log and a screenshot from the page when using this code:
```
<script type='text/javascript'>
    EJS_player = '#game';
    EJS_core = 'nes';
    EJS_oldCores = true;
    EJS_VirtualGamepadSettings = [
        {
        type: "button",
        text: "Y",
        id: "y",
        location: "right",
        left: 40,
        bold: true,
        input_value: 9
        },
        {
        type: "button",
        text: "X",
        id: "X",
        location: "right",
        top: 40,
        bold: true,
        input_value: 1
        },
        {
        type: "button",
        text: "B",
        id: "b",
        location: "right",
        left: 81,
        top: 40,
        bold: true,
        input_value: 8
        },
        {
        type: "button",
        text: "A",
        id: "a",
        location: "right",
        left: 40,
        top: 80,
        bold: true,
        input_value: 0
        },
        {
        type: "button",
        text: "Start",
        id: "start",
        location: "center",
        left: 60,
        fontSize: 15,
        block: true,
        input_value: 3
        },
        {
        type: "button",
        text: "Select",
        id: "select",
        location: "center",
        left: -5,
        fontSize: 15,
        block: true,
        input_value: 2
        }
]
    EJS_gameUrl = '';
    EJS_pathtodata = ';
    EJS_gameID = 1;
</script>
<script src="/loader.js"></script>
```
I intentionally left out the Zone or dpad settings because adding either one of those makes the emulator not load at all. 

[emulatorjs.log](https://github.com/EmulatorJS/EmulatorJS/files/11648382/emulatorjs.log)

![IMG_3720](https://github.com/EmulatorJS/EmulatorJS/assets/135175291/06e57eba-bb14-48b5-a4ee-aaf61ef23f96)

**Reply by ethanaobrien** [reply]
Could you do these things

1. Could you update the emulatorjs files, I made some changed yesterday
2. Set `EJS_DEBUG_XX` to `true`
3. Could you then send a log and the code when you add the gamepad controls that make it not work. I understand that they make the page not work, which is why I need the code and the log.

**Reply by S-Coding23** [reply]
Thank you. I updated the emulatorjs files then used the below code for the page, and attached
 the console logs. 
```
<script type='text/javascript'>
    EJS_DEBUG_XX = true; 
    EJS_player = '#game';
    EJS_core = 'nes';
    EJS_oldCores = true;
    EJS_VirtualGamepadSettings = [
        {
        type: "button",
        text: "Y",
        id: "y",
        location: "right",
        left: 40,
        bold: true,
        input_value: 9
        },
        {
        type: "button",
        text: "X",
        id: "X",
        location: "right",
        top: 40,
        bold: true,
        input_value: 1
        },
        {
        type: "button",
        text: "B",
        id: "b",
        location: "right",
        left: 81,
        top: 40,
        bold: true,
        input_value: 8
        },
        {
        type: "button",
        text: "A",
        id: "a",
        location: "right",
        left: 40,
        top: 80,
        bold: true,
        input_value: 0
        },
        {
        type: "dpad"
        location: "left",
        left: "50%",
        right: "50%",
        joystickInput: false,
        inputValues: [4, 5, 6, 7]
        },
        {
        type: "button",
        text: "Start",
        id: "start",
        location: "center",
        left: 60,
        fontSize: 15,
        block: true,
        input_value: 3
        },
        {
        type: "button",
        text: "Select",
        id: "select",
        location: "center",
        left: -5,
        fontSize: 15,
        block: true,
        input_value: 2
        }
]
    EJS_gameUrl = '';
    EJS_pathtodata = ';
    EJS_gameID = 1;
</script>
<script src="/loader.js"></script>
```
[emulatorjsDebug.log](https://github.com/EmulatorJS/EmulatorJS/files/11657978/emulatorjsDebug.log)

**Reply by ethanaobrien** [reply]
![Screenshot_20230605_174418_GitHub.jpg](https://github.com/EmulatorJS/EmulatorJS/assets/77750390/43d35239-0394-48f2-9004-7a81edb919a5)

Looks like you might just be missing a comma. Try adding one and see if that fixes it

**Reply by S-Coding23** [reply]
And thats why you dont copy and paste....
That was it. Thank you!

**Reply by ethanaobrien** [reply]
Yeah! No problem! Just wondering, was this the code from the website that was missing a comma?

**Reply by S-Coding23** [reply]
It is. Just the dpad entry is missing the comma.
